keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1490 @@
+    the IPv4 case without needing `route_localnet` in that case.) By
+    the time the packet has reached the network stack though, it is
+    probably too late to change the source IP? (OTOH, using eBPF to
+    intercept the `connect()` call would probably work.)
+
+  - It is possible that using the nftables `fwd` command from a rule
+    on a table in the `netdev` family would let us move the packet
+    around in a way that bypassed the kernel checks that otherwise
+    make this not work?
+
+  - We could resurrect enough of the userspace proxy to use it just
+    for localhost NodePort connections... :vomiting_face:
+
+      - (Alternatively: having a listening port on localhost, but then
+        after accepting a connection, using eBPF to splice it directly
+        to the destination server socket.)
+
+It's not clear what the use cases for localhost NodePorts are. Tim's
+[initial implementation of NodePorts in the iptables proxy] set
+`route_localnet`, so localhost nodeports seem to have been considered
+an expected feature already at that point.
+
+  - In kubernetes #90236, Tim commented that one use case is
+    [connecting to a docker registry without needing to set up
+    certificates]; docker special-cases `localhost` specifically as
+    not needing TLS. (Lars also mentions in #92938 that this is "[the
+    only use-case I have heard]".)
+
+  - In other cases, `localhost` was used purely for simplicity, and it
+    is possible to rewrite the code to [use the node IP instead] (eg,
+    passed via the downward API).
+
+[the client explicitly set a non-localhost source IP]: https://github.com/kubernetes/kubernetes/issues/90236#issuecomment-622355484
+[initial implementation of NodePorts in the iptables proxy]: https://github.com/kubernetes/kubernetes/pull/12764
+[connecting to a docker registry without needing to set up certificates]: https://github.com/kubernetes/kubernetes/issues/90236#issuecomment-622142806
+[the only use-case I have heard]: https://github.com/kubernetes/kubernetes/pull/92938#issuecomment-742985382
+[use the node IP instead]: https://github.com/contiv/vpp/pull/1434
+
+#### NodePort Addresses
+
+In addition to the localhost issue, iptables kube-proxy defaults to
+accepting NodePort connections on all local IPs, which has effects
+varying from intended-but-unexpected ("why can people connect to
+NodePort services from the management network?") to clearly-just-wrong
+("why can people connect to NodePort services on LoadBalancer IPs?")
+
+The nftables proxy should default to only opening NodePorts on a
+single interface, probably the interface with the default route by
+default. (Ideally, you really want it to accept NodePorts on the
+interface that holds the route to the cloud load balancers, but we
+don't necessarily know what that is ahead of time.) Admins can use
+`--nodeport-addresses` to override this.
+
+#### Behavior of service IPs
+
+Tim has suggested that service IPs should reject connections on ports
```

Comment:
```suggestion
Tim Hockin has suggested that service IPs should reject connections on ports
```
We have multiple Tims and they don't support selecting based on annotations.

Reply:
There's sort of an implied `<<UNRESOLVED>>` around the entire "Changes from iptables" section; we'll want to resolve a lot of details here before merging. (And to the extent that we don't, yes, I'll fix things up before then.)